File: lib/haveapi/cli/authentication/basic.rb
module HaveAPI::CLI::Authentication
  class Basic < Base
    register :basic

    def options(opts)
      opts.on('--username USER', 'User name') do |u|
        @user = u
      end

      opts.on('--password PASSWORD', 'Password') do |p|
        @password = p
      end
    end

    def validate
      @user ||= ask('User name: ') { |q| q.default = nil }

      @password ||= ask('Password: ') do |q|
        q.default = nil
        q.echo = false
      end
    end

    def authenticate
      @communicator.authenticate(:basic, {user: @user, password: @password})
    end
  end
end

File: lib/haveapi/authentication/basic.ex
defmodule HaveAPI.Authentication.Basic do
  @callback find_user(%Plug.Conn{}, String.t, String.t) :: any

  defmacro __using__(_opts) do
    quote do
      @behaviour unquote(__MODULE__)

      def name, do: :basic

      def describe(_ctx) do
        %{
          description: "Authentication using HTTP basic. Username and password is passed " <>
                       "via HTTP header. Its use is forbidden from web browsers."
        }
      end

      def authenticate(conn) do
        header = Enum.find(
          conn.req_headers,
          fn {k,v} -> k == "authorization" end
        )

        if header do
          {_, auth} = header
          [_, auth] = String.split(auth, ~r{\s}, trim: true)
          [user, password] = auth |> Base.decode64! |> String.split(":")

          find_user(conn, user, password)

        else
          nil
        end

      rescue ArgumentError ->
        nil
      end
    end
  end
end
